types/prefs/list.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package prefs
+
+import (
+	"net/netip"
+	"slices"
+	"time"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"golang.org/x/exp/constraints"
+	"tailscale.com/types/opt"
+	"tailscale.com/types/ptr"
+	"tailscale.com/types/views"
+)
+
+// BasicType is a constraint that allows types whose underlying type is a predeclared
+// boolean, numeric, or string type.
+type BasicType interface {
+	~bool | constraints.Integer | constraints.Float | constraints.Complex | ~string
+}
+
+// ImmutableType is a constraint that allows [BasicType]s and certain well-known immutable types.
+type ImmutableType interface {
+	BasicType | time.Time | netip.Addr | netip.Prefix | netip.AddrPort
+}
+
+// List is a preference type that holds zero or more values of an [ImmutableType] T.
+type List[T ImmutableType] struct {
+	preference[[]T]
+}
+
+// ListOf returns a [List] configured with the specified value and [Options].
+func ListOf[T ImmutableType](v []T, opts ...Options) List[T] {
+	return List[T]{preferenceOf(opt.ValueOf(v), opts...)}
+}
+
+// ListWithOpts returns an unconfigured [List] with the specified [Options].
+func ListWithOpts[T ImmutableType](opts ...Options) List[T] {
+	return List[T]{preferenceOf(opt.Value[[]T]{}, opts...)}
+}
+
+// View returns a read-only view of l.
+func (l *List[T]) View() ListView[T] {
+	return ListView[T]{l}
+}
+
+// Clone returns a copy of l that aliases no memory with l.
+func (l *List[T]) Clone() *List[T] {
+	res := ptr.To(*l)
+	if v, ok := l.s.Value.GetOk(); ok {
+		res.s.Value.Set(append(v[:0:0], v...))
+	}
+	return res
+}
+
+// Equal reports whether l and l2 are equal.
+func (l List[T]) Equal(l2 List[T]) bool {
+	if l.s.Metadata != l2.s.Metadata {
+		return false
+	}
+	v1, ok1 := l.s.Value.GetOk()
+	v2, ok2 := l2.s.Value.GetOk()
+	if ok1 != ok2 {
+		return false
+	}
+	return !ok1 || slices.Equal(v1, v2)
+}
+
+// ListView is a read-only view of a [List].
+type ListView[T ImmutableType] struct {
+	// ж is the underlying mutable value, named with a hard-to-type
+	// character that looks pointy like a pointer.
+	// It is named distinctively to make you think of how dangerous it is to escape
+	// to callers. You must not let callers be able to mutate it.
+	ж *List[T]
+}
+
+// Valid reports whether the underlying [List] is non-nil.
+func (lv ListView[T]) Valid() bool {
+	return lv.ж != nil
+}
+
+// AsStruct returns a clone of the [List] which aliases no memory with the original.
+func (lv ListView[T]) AsStruct() *List[T] {
```

Comment:
`ListView` implements [`views.StructView`](https://github.com/tailscale/tailscale/blob/ee976ad704980e20ec36c6aaaad0a2ce5b30b3d5/types/views/views.go#L107):
```
// StructView represents the corresponding StructView of a Viewable. The concrete types are
// typically generated by tailscale.com/cmd/viewer.
type StructView[T any] interface {
	// Valid reports whether the underlying Viewable is nil.
	Valid() bool
	// AsStruct returns a deep-copy of the underlying value.
	// It returns nil, if Valid() is false.
	AsStruct() T
}
```
I'll a comment to clarify this.

Reply:
Does it need to implement that interface? Other slice types in that package seem to implement `AsSlice` methods instead https://github.com/tailscale/tailscale/blob/ee976ad704980e20ec36c6aaaad0a2ce5b30b3d5/types/views/views.go#L197-L200

Reply:
Yes and yes. A `prefs.List` and a `prefs.ListView` represent a preference and a read-only view of a preference.
The preference's _value type_ is a slice. Therefore, for example, `ListView.Value` returns a `view.Slice`:
https://github.com/tailscale/tailscale/blob/5be7fbbedaf124760798ddd35b08ad25dfbfd804/types/prefs/list.go#L100-L104

But a `ListView` itself is a read-only view of a struct, and it has to implement `views.StructView` both from the semantics perspective and in order to be [a valid container type](https://github.com/tailscale/tailscale/pull/12809).

Reply:
@awly: what I think we should eventually do, which may or may not be enough for the `prefs` package specifically but would most certainly be useful in general, is to further improve the `cmd/viewer` code generation so that the view types we've got in the `prefs` package (e.g., `ItemView`, `ListView`, `StructListView`, and `StructMapView`) could be generated automatically from the corresponding struct types (e.g., `prefs.Item`, `prefs.List`, `prefs.StructList`, and `prefs.StructMap`). The generated views would expose the underlying type's exported methods that have value receivers and wrap their output parameters as read-only views when necessary. Ideally, we would also generate functions like [`StructListViewOf`](https://github.com/tailscale/tailscale/blob/5be7fbbedaf124760798ddd35b08ad25dfbfd804/types/prefs/struct_list.go#L92-L96) for the generated views, but these are small enough to be hand-written.

I've filed an issue to track this effort: https://github.com/tailscale/tailscale/issues/13191
But having worked on `cmd/viewer`, I think we should probably refactor and clean it up sometime before introducing more complex features like this to it.